core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -4139,6 +4139,147 @@ class KafkaApis(val requestChannel: RequestChannel,
+        })
+      } else {
+        val name = topicNames.get(acknowledgeTopic.topicId)
+        acknowledgeTopic.partitions.forEach((acknowledgePartition: ShareAcknowledgeRequestData.AcknowledgePartition) =>
+          requestPartitionAcknowledgements +=
+            new TopicIdPartition(
+              acknowledgeTopic.topicId,
+              new TopicPartition(name, acknowledgePartition.partitionIndex)) ->
+              acknowledgePartition
+        )
+      }
+    })
+    val authorizedTopics = authHelper.filterByAuthorized(
+      request.context,
+      READ,
+      TOPIC,
+      requestPartitionAcknowledgements
+    )(_._1.topicPartition.topic)
+
+    requestPartitionAcknowledgements.foreach {
+      case (topicIdPartition : TopicIdPartition, acknowledgePartition : ShareAcknowledgeRequestData.AcknowledgePartition) =>
+        if (!authorizedTopics.contains(topicIdPartition.topic))
+          erroneous += topicIdPartition ->
+            ShareAcknowledgeResponse.partitionResponse(topicIdPartition, Errors.TOPIC_AUTHORIZATION_FAILED)
+        else if (!metadataCache.contains(topicIdPartition.topicPartition))
+          erroneous += topicIdPartition ->
+            ShareAcknowledgeResponse.partitionResponse(topicIdPartition, Errors.UNKNOWN_TOPIC_OR_PARTITION)
+        else
+          interesting += topicIdPartition -> acknowledgePartition
+    }
+
+    // the callback for processing a share acknowledge response, invoked before throttling
+    def processResponseCallback(
+                                 responseAcknowledgeData: Map[TopicIdPartition,
+                                   ShareAcknowledgeResponseData.PartitionData]
+                               ): Unit = {
+      val partitions = new util.LinkedHashMap[TopicIdPartition, ShareAcknowledgeResponseData.PartitionData]
+      val nodeEndpoints = new mutable.HashMap[Int, Node]
+      responseAcknowledgeData.foreach { case(tp, data) =>
+        val partitionData = new ShareAcknowledgeResponseData.PartitionData()
+          .setPartitionIndex(tp.partition)
+          .setErrorCode(data.errorCode())
+
+        data.errorCode() match {
+          case errCode if errCode == Errors.NOT_LEADER_OR_FOLLOWER.code | errCode == Errors.FENCED_LEADER_EPOCH.code =>
+            val leaderNode = getCurrentLeader(tp.topicPartition(), request.context.listenerName)
+            leaderNode.node.foreach { node =>
+              nodeEndpoints.put(node.id(), node)
+            }
+            partitionData.currentLeader()
+              .setLeaderId(leaderNode.leaderId)
+              .setLeaderEpoch(leaderNode.leaderEpoch)
+          case _ =>
+        }
+
+        partitions.put(tp, partitionData)
+      }
+      erroneous.foreach { case (tp, data) => partitions.put(tp, data) }
+
+      var unconvertedShareAcknowledgeResponse: ShareAcknowledgeResponse = null
+
+      def createResponse(): ShareAcknowledgeResponse = {
+        // Down-convert messages for each partition if required
```

Comment:
Thanks for the review ! I have removed the comment. I don't think we need down conversion here a all, because we don't have any record information present in ShareAcknowledgeResponseData.